framework/Optimizers/Conjugate.py

```diff
@@ -0,0 +1,421 @@
+# Copyright 2017 Battelle Energy Alliance, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+  This module contains the Conjugate Gradient Optimization strategy
+
+  Created on Sept 17, 2019
+  @ author: ZHOUJ2
+"""
+#for future compatibility with Python 3--------------------------------------------------------------
+from __future__ import division, print_function, unicode_literals, absolute_import
+import warnings
+warnings.simplefilter('default',DeprecationWarning)
+#End compatibility block for Python 3----------------------------------------------------------------
+
+#External Modules------------------------------------------------------------------------------------
+import sys
+import os
+import copy
+import numpy as np
+from numpy import linalg as LA
+import scipy
+from scipy.optimize import minpack2
+#External Modules End--------------------------------------------------------------------------------
+
+#Internal Modules------------------------------------------------------------------------------------
+from .SPSA import SPSA
+from utils import mathUtils,randomUtils
+
+#Internal Modules End--------------------------------------------------------------------------------
+
+class ConjugateGradient(SPSA):
+  """
+    Finite Difference Gradient Optimizer
+    This class currently inherits from the SPSA (since most of the gradient based machinery is there).
+    TODO: Move the SPSA machinery here (and GradientBasedOptimizer) and make the SPSA just take care of the
+    random perturbation approach
+  """
+  ##########################
+  # Initialization Methods #
+  ##########################
+  def __init__(self):
+    """
+      Default Constructor
+    """
+    SPSA.__init__(self)
+
+
+  def localInputAndChecks(self, xmlNode):
+    """
+      Local method for additional reading.
+      @ In, xmlNode, xml.etree.ElementTree.Element, Xml element node
+      @ Out, None
+    """
+    SPSA.localInputAndChecks(self, xmlNode)
+    # need extra eval for central Diff, using boolean in math
+    self.useGradHist = False
+    self.paramDict['pertSingleGrad'] = (1 + self.useCentralDiff) * len(self.fullOptVars)
+    self.gradDict['pertNeeded'] = self.gradDict['numIterForAve'] * (self.paramDict['pertSingleGrad']+1)
+    if self.useCentralDiff:
```

Comment:
isn't this common to all the Gradient Based?

Reply:
For SPSA it's a fixed value